File: with-golang-gorm/lakebase/lakebase.go
// Package lakebase fetches short-lived DB tokens from Databricks and returns a connection string.
package lakebase

import (
	"bytes"
	"encoding/base64"
	"encoding/json"
	"fmt"
	"net/http"
	"net/url"
	"os"
)

// GetConnectionString returns a PostgreSQL connection string using a token from the Databricks API.
func GetConnectionString() (string, error) {
	host := os.Getenv("DATABRICKS_HOST")
	clientID := os.Getenv("DATABRICKS_CLIENT_ID")
	clientSecret := os.Getenv("DATABRICKS_CLIENT_SECRET")
	endpoint := os.Getenv("LAKEBASE_ENDPOINT")
	dbHost := os.Getenv("LAKEBASE_HOST")
	dbPort := os.Getenv("LAKEBASE_PORT")
	if dbPort == "" {
		dbPort = "5432"
	}
	dbName := os.Getenv("LAKEBASE_DATABASE")
	if dbName == "" {
		dbName = "databricks_postgres"
	}
	if host == "" || clientID == "" || clientSecret == "" || endpoint == "" || dbHost == "" {
		return "", fmt.Errorf("set DATABRICKS_HOST, DATABRICKS_CLIENT_ID, DATABRICKS_CLIENT_SECRET, LAKEBASE_ENDPOINT, LAKEBASE_HOST")
	}

	creds := base64.StdEncoding.EncodeToString([]byte(clientID + ":" + clientSecret))
	oidcReq, _ := http.NewRequest("POST", "https://"+host+"/oidc/v1/token", bytes.NewBufferString("grant_type=client_credentials&scope=all-apis"))
	oidcReq.Header.Set("Content-Type", "application/x-www-form-urlencoded")
	oidcReq.Header.Set("Authorization", "Basic "+creds)
	oidcRes, err := http.DefaultClient.Do(oidcReq)
	if err != nil {
		return "", err
	}
	defer oidcRes.Body.Close()
	var oidc struct {
		AccessToken string `json:"access_token"`
	}
	if err := json.NewDecoder(oidcRes.Body).Decode(&oidc); err != nil {
		return "", err
	}

	body, _ := json.Marshal(map[string]string{"endpoint": endpoint})
	dbReq, _ := http.NewRequest("POST", "https://"+host+"/api/2.0/postgres/credentials", bytes.NewReader(body))
	dbReq.Header.Set("Content-Type", "application/json")
	dbReq.Header.Set("Authorization", "Bearer "+oidc.AccessToken)
	dbRes, err := http.DefaultClient.Do(dbReq)
	if err != nil {
		return "", err
	}
	defer dbRes.Body.Close()
	var cred struct {
		Token string `json:"token"`
	}
	if err := json.NewDecoder(dbRes.Body).Decode(&cred); err != nil {
		return "", err
	}

	password := url.QueryEscape(cred.Token)
	user := url.QueryEscape(clientID)
	return fmt.Sprintf("postgresql://%s:%s@%s:%s/%s?sslmode=require", user, password, dbHost, dbPort, dbName), nil
}

File: with-rust-tokio-postgres/src/lakebase.rs
// Fetches short-lived DB token from Databricks and returns a PostgreSQL connection string.
use serde::Deserialize;
use std::env;

#[derive(Deserialize)]
struct OidcResponse {
    access_token: String,
}

#[derive(Deserialize)]
struct CredResponse {
    token: String,
}

pub async fn get_connection_string() -> Result<String, Box<dyn std::error::Error + Send + Sync>> {
    let host = env::var("DATABRICKS_HOST")?;
    let client_id = env::var("DATABRICKS_CLIENT_ID")?;
    let client_secret = env::var("DATABRICKS_CLIENT_SECRET")?;
    let endpoint = env::var("LAKEBASE_ENDPOINT")?;
    let db_host = env::var("LAKEBASE_HOST")?;
    let db_port = env::var("LAKEBASE_PORT").unwrap_or_else(|_| "5432".to_string());
    let db_name = env::var("LAKEBASE_DATABASE").unwrap_or_else(|_| "databricks_postgres".to_string());

    let http = reqwest::Client::new();
    let credentials = base64::engine::general_purpose::STANDARD.encode(format!("{}:{}", client_id, client_secret));

    let oidc_res = http
        .post(format!("https://{}/oidc/v1/token", host))
        .header("Content-Type", "application/x-www-form-urlencoded")
        .header("Authorization", format!("Basic {}", credentials))
        .body("grant_type=client_credentials&scope=all-apis")
        .send()
        .await?
        .json::<OidcResponse>()
        .await?;

    let cred_res = http
        .post(format!("https://{}/api/2.0/postgres/credentials", host))
        .header("Content-Type", "application/json")
        .header("Authorization", format!("Bearer {}", oidc_res.access_token))
        .json(&serde_json::json!({ "endpoint": endpoint }))
        .send()
        .await?
        .json::<CredResponse>()
        .await?;

    let user = urlencoding::encode(&client_id);
    let password = urlencoding::encode(&cred_res.token);
    Ok(format!(
        "postgresql://{}:{}@{}:{}/{}?sslmode=require",
        user, password, db_host, db_port, db_name
    ))
}
